tests/create_task_output.yml

```diff
@@ -0,0 +1,47 @@
+name: Create Tes Task Job With Output
+description: Job to create a new TES Task with output
```

Comment:
Which new feature/schema is this test being added for?

Reply:
I didn't check the other tests, but if there's no test with outputs, and outputs - if available - has required parameters, then I think the test makes sense.

See discussion in next comment for details.

Reply:
Yes, you are right. Currently, there are no tests to check the outputs specifically.
Keeping v1.0.0 in mind, we can add a test having all required properties (3 in number). For v1.1.0, there are 2 required properties. I will log an issue for it and add the new tests.

Just wanted to discuss, how should we handle the test file names? Both are exactly same, just the version is different. Perhaps, for now we can keep the version in the file name. And when we separate the tests and executors, we can have test folders specific to versions.

Reply:
In that case, yes, please add a test for outputs for 1.0.0. And if there are other such cases (i.e., non-required properties that do have required child properties, if provided), then please add tests for these, too, if not already available.

As for the names: I think you should try and disentangle two roles here, that of the developer of the test runner and that of the developer of _test cases_ (now you do both, but the idea is that sooner or later, the second role will be done by other people, both for TES and maybe other specs). And I think that as a result of the confusion of these two roles, we are currently using too many different things for the same purpose (e.g., we have the test file path/name, the `name` field, and we put a unique tag for each test case).

What I would do, is to actually remove the `name` field altogether. It is difficult to come up with naming conventions that account for all possible cases, it is very difficult to enforce naming conventions, and you make it difficult for yourself either enforcing unique names (extra code, and then what? Erroring out?) or _not_ enforcing unique names (how do you report the results of a given test if you can't uniquely map it back). And last but not least, you put an unnecessary cognitive burden on the people writing tests. And for what benefit? That people hopefully use semantic names for tests? But they will still write a description (hopefully) and ... here's the thing ... give their test file a name - which, together with its path, is _guaranteed_ to be unique (note that you could also use a single path/filename as an argument rather than a directory of tests - this is how, e.g., `pytest` works as well; so no need for unique tags for a given test)!

So, I would say, for the runner, don't worry about the naming or the file subdirectories in which test developers organize the tests: Instead, get all the valid tests in the target directory and subdirectories, filter by tags, and use the filename with path as their unique identifiers. Users will then need to dig into the tests (perhaps a `--verbose` option would also print out the test description in the report? If it does not already do so), but what can you do? The test runner can't ever be generic _and_ understand the semantics of a set of tests at the same time. To make it easy for users (people running the tests) to understand what a given test does by describing it will in the field that is dedicated for it.

Likewise, with the versions: _Every_ test is specific to a set of versions. Anything in the API can change, at any time, so a test that _seems_ to be "generic" now, may not be so in TES 2.0.0, or TES 7.3.1. I see two ways of how to deal with it:
1. You treat versions as nothing special and tell test developers that they must use tags to support versioned APIs. And then don't have any version-specific code in your runner. Of course this option puts a bit of a burden on test developers if there are a lot of versions (who then need to enumerate them in the tags of each test) and also makes the tests more messy.
2. You _do_ recognize that versions are basically integral to API tests and build in proper support for them. In that case, to minimize code/test repetition, I would ask test developers to provide a mapping of tests (by path/filename) for each version. Then you can remove the version tags. And by default you can always use the mapping for the _lastest_ version (by SemVer ordering rules). This is what I would do, although option (1) seems fair to me as well.

Speaking of tags (and now moving to the test developer role), I find the `logical` and `functional` tags to be a bit ambiguous / not self evident. And again, I think that there are some integral things going on again that you could consider adding proper support for (outside of tags).

Basically, there are two types of tests, with the first type having itself two subtypes:
1. The service does what it is supposed to do. These are currently tagges as `functional`. Out of these, some may be (1a) required by the specification, while others are (1b) optional.
2. The service accepts what it is supposed to accept and returns what it is supposed to return. These tests are currently tagged as `logical`. I may be wrong, but I think all of these are, by definition, required by the specification.

In the end, we would probably be able to give a service a sort of "green badge" if - and only if! - all tests (for the correct schema version) that fill into categories (1a) and (2) succeed. Then, for optional functionalities, we would need to parse the report and see if a given service supports it or not.

There are some ways how to implement that:
1. Add two boolean flags: `required` and `functional`. The problem here is that optional non-functional tests probably don't exist, and also this setup would not be very extensible. `required` is probably truly boolean, it either is or isn't. But whether something is a functional test, or logical/validation or maybe something else is not so clear, as `functional` is not a true antonym of `logical`, and `non-functional` is a weird and unhelpful label. So maybe let's not do that.
2. Enumerate `validation_only`, `functional_required` and `functional_optional`. That would probably work and is extensible, but is difficult to parse, especially if it _is_ extended in the future.
3. So perhaps from the test runner side we don't really care if a test is `logical` or `functional` to give that green badge. We don't need to know whatever a test developer means when they label a test like that. It may mean something to _them_ and hopefully also to the user, but not to the test runner. So why not add a boolean flag `required` to the test schema, and make use of that to decide whether a test run passed or failed. And for further classifications, we still have the `tag` field.

For example, for our situtation, I would simply use a single tag `schema_validation_only` (more descriptive than `logical`). I don't really see the need/use for any other labels than that. If a user really _only_ wanted to run "functional" tests (which is a bit weird, but okay), and we wanted to support that, we could implement a negative tag filter for the runner.

Finally, I would only allow letters, digits and underscores in tag names; it's not much of a restriction for test developers, but it makes parsing from the command line much easier. If you want to filter by multiple tags, being able to do `--include tag1,tag2` is much more user-friendly than having to deal with quotes and backslash escpates and whatnot.

Okay, that became a very very long comment. Wasn't intended, so sorry about that.

tl;dr

1. Remove the `name` field from test case schema and tags for individual test cases. Use the path/filename instead (already unique!). Allow users to specify paths to one or more test cases or directories of tests and run all tests that are found, filtered by tags, if available.
2. Remove version tags. Bake in support for versioning by asking test developers to provide mappings from versions to test case paths for each version. Add a functionality to discover available versions for the set of test cases (by parsing that dictionary). Use the most recent version as default if not version is supplied.
3. Provide `required` field in test case schema to decide if a test run passed or failed.
4. Give the user the ability to do a negative selection on tags (`--include-tags` vs `--exclude_tags`). Remove excessive tags from the available test cases, e.g., keep only one of `logical` (better: `schema_validation_only`) or `functional` tags. With positive and negative tag filtering, this will be powerful enough.

Those are my suggestions anyway. If you agree with any of these, please create issues for those, otherwise ignore :)

Reply:
Thank you @uniqueg for this detailed suggestion. I understood that these things might convolute the code/logic in future and it is best to simplify them as you mentioned.

For names: Mapping the file paths/names as unique identifiers will indeed serve as an easier alternative to defining unique names as done currently in every test.
We can add the test description in the report without the `--verbose` flag. They should not be lengthy and we will always be able to understand the purpose of the test.

In versions, I understand that with multiple versions, the test developers would need to add new versions in every test. But, the test developers are closer to the service and they should be able to manage this. 
Adding support on the runner side will complicate the division of responsibilities. For every test-version mapping change in the "service" API, the change would be on the "runner" (unrelated to any runner change). This will also create a direct dependency on the runner team.

In every way, we have to store the test to version mapping. The test developers (who deal with both, tests and versions) will be the better choice to store them than the test runners (who only deal with versions). We can maybe add this in future if we think that the version tags are becoming unmanageable.

Thinking about the optional `functional` tests, all the tests are based on the API specification. If a test is written, it should _pass_ for _all_ the API services. For an optional test, it suggests that it was _not_ defined in the specification, for which we don't need to write that test. I think all tests are required, irrespective of being `logical` or `functional.`

The negative tag filter is a great idea. The single tag `- schema_validation_only` can be efficiently used then.

The CLI arg for tag can be upgraded as you suggested. Adding constraints on the tags would allow this implementation.